src/Functions/countMatches.h

```diff
@@ -0,0 +1,120 @@
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionCountMatches<CountMatchesBase>>(); }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 2; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (!isStringOrFixedString(arguments[1].type))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Illegal type {} of second argument (pattern) of function {}. Must be String/FixedString.",
+                arguments[1].type->getName(), getName());
+        if (!isStringOrFixedString(arguments[0].type))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Illegal type {} of first argument (haystack) of function {}. Must be String/FixedString.",
+                arguments[0].type->getName(), getName());
+        const auto * column = arguments[1].column.get();
+        if (!column || !checkAndGetColumnConstStringOrFixedString(column))
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                "The second argument of function {} should be a constant string with the pattern",
+                getName());
+
+        return std::make_shared<DataTypeUInt64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
+    {
+        const ColumnConst * column_pattern = checkAndGetColumnConstStringOrFixedString(arguments[1].column.get());
+        Regexps::Pool::Pointer re = Regexps::get<false /* like */, true /* is_no_capture */, CountMatchesBase::case_insensitive>(column_pattern->getValue<String>());
+        OptimizedRegularExpression::MatchVec matches;
+
+        const IColumn * column_haystack = arguments[0].column.get();
+
+        if (const ColumnString * col_str = checkAndGetColumn<ColumnString>(column_haystack))
+        {
+            auto result_column = ColumnUInt64::create();
+
+            const ColumnString::Chars & src_chars = col_str->getChars();
+            const ColumnString::Offsets & src_offsets = col_str->getOffsets();
+
+            ColumnUInt64::Container & vec_res = result_column->getData();
+            vec_res.resize(input_rows_count);
+
+            size_t size = src_offsets.size();
+            ColumnString::Offset current_src_offset = 0;
+
+            for (size_t i = 0; i < size; ++i)
+            {
+                Pos pos = reinterpret_cast<Pos>(&src_chars[current_src_offset]);
+                current_src_offset = src_offsets[i];
+                Pos end = reinterpret_cast<Pos>(&src_chars[current_src_offset]) - 1;
+
+                StringRef str(pos, end - pos);
+                vec_res[i] = countMatches(str, re, matches);
+            }
+
+            return result_column;
+        }
+        else if (const ColumnConst * col_const_str = checkAndGetColumnConstStringOrFixedString(column_haystack))
```

Comment:
Should we use `useDefaultImplementationForConstants`? Like here: https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/FunctionsMultiStringFuzzySearch.h#L49

Reply:
>Should we use useDefaultImplementationForConstants?

But countMatches() handle it by itself, so no need in default implementation.

>Like here: https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/FunctionsMultiStringFuzzySearch.h#L49

But here it is needed, since the haystack can be null, and this is not handled